devmand/gateway/src/devmand/channels/cli/SshSessionAsync.cpp

```diff
@@ -25,48 +25,93 @@ using devmand::channels::cli::sshsession::SshSession;
 using devmand::channels::cli::sshsession::SshSessionAsync;
 using std::lock_guard;
 using std::unique_lock;
+using namespace std::chrono_literals;
 
 SshSessionAsync::SshSessionAsync(
     string _id,
-    shared_ptr<IOThreadPoolExecutor> _executor)
-    : executor(_executor), session(_id), reading(false) {}
+    shared_ptr<folly::Executor> _executor)
+    : id(_id),
+      executor(_executor),
+      serialExecutor(SerialExecutor::create(
+          folly::Executor::getKeepAliveToken(_executor.get()))),
+      session(_id),
+      reading(false),
+      callbackFinished(false),
+      matchingExpectedOutput(false) {}
+
+static const int EVENT_FINISH = 9999;
 
 SshSessionAsync::~SshSessionAsync() {
-  if (this->sessionEvent != nullptr &&
-      event_get_base(this->sessionEvent) != nullptr) {
-    event_free(this->sessionEvent);
+  MLOG(MDEBUG) << "~SshSessionAsync started";
+
+  // Let the NIO callback finish by injecting artificial event and waiting for
+  // callback to finish
+  if (sessionEvent != nullptr && event_get_base(sessionEvent) != nullptr) {
+    event_active(this->sessionEvent, EVENT_FINISH, -1);
+    while (!callbackFinished) {
+      std::this_thread::sleep_for(100ms);
```

Comment:
is the problem here the callback is the event loop using a deleted obj? can we instead use a weak ptr?